lib/Model/FirewallRule.php

```diff
@@ -1,867 +1,493 @@
-            throw new \InvalidArgumentException(
-                sprintf(
-                    "Invalid value for 'direction', must be one of '%s'",
-                    implode("', '", $allowed_values)
-                )
-            );
-        }
-        $this->container['direction'] = $direction;
+    private $sourcePortStart;
 
-        return $this;
-    }
+    /**
+     * @var string|float|null
+     */
+    private $sourcePortEnd;
 
     /**
-     * Gets action
-     * @return string
+     * @var string|null
      */
-    public function getAction()
-    {
-        return $this->container['action'];
-    }
+    private $comment;
 
     /**
-     * Sets action
-     * @param string $action Action to take if the rule conditions are met.
-     * @return $this
+     * Constructor
+     * @param mixed[] $data
      */
-    public function setAction($action)
+    public function __construct(array $data = null)
     {
-        $allowed_values = $this->getActionAllowableValues();
-        if (!in_array($action, $allowed_values)) {
-            throw new \InvalidArgumentException(
-                sprintf(
-                    "Invalid value for 'action', must be one of '%s'",
-                    implode("', '", $allowed_values)
-                )
-            );
-        }
-        $this->container['action'] = $action;
-
-        return $this;
+        $this->setDirection($data['direction'] ?? null);
+        $this->setAction($data['action'] ?? null);
+        $this->setPosition($data['position'] ?? null);
+        $this->setFamily($data['family'] ?? null);
+        $this->setProtocol($data['protocol'] ?? null);
+        $this->setIcmpType($data['icmp_type'] ?? null);
+        $this->setDestinationAddressStart($data['destination_address_start'] ?? null);
+        $this->setDestinationAddressEnd($data['destination_address_end'] ?? null);
+        $this->setDestinationPortStart($data['destination_port_start'] ?? null);
+        $this->setDestinationPortEnd($data['destination_port_end'] ?? null);
+        $this->setSourceAddressStart($data['source_address_start'] ?? null);
+        $this->setSourceAddressEnd($data['source_address_end'] ?? null);
+        $this->setSourcePortStart($data['source_port_start'] ?? null);
+        $this->setSourcePortEnd($data['source_port_end'] ?? null);
+        $this->setComment($data['comment'] ?? null);
     }
 
     /**
-     * Gets position
-     * @return float
+     * @return string
      */
-    public function getPosition()
+    public function getDirection(): string
     {
-        return $this->container['position'];
+        return $this->direction;
     }
 
     /**
-     * Sets position
-     * @param float $position Add the firewall rule to this position in the server's firewall list.
-     * @return $this
+     * @param string|null $direction
+     * @return FirewallRule
      */
-    public function setPosition($position)
+    public function setDirection(?string $direction): FirewallRule
```

Comment:
Seems to be unable to receive anything or always defaults to null for some reason. Tried:

```
$firewall_rule->setDirection(\Upcloud\ApiClient\Model\FirewallRule::DIRECTION_OUT);
$firewall_rule->setDirection('out');
```

Reply:
@PopoSensei could you also share a code where you're creating a `$firewall_rule`? 

There's definitely something wrong on our side, but it would be great to see how exactly you're using this Model before we make fixes.

Reply:
In all simplicity this is what I tried for now:

```
<?php
require_once(__DIR__ . '/vendor/autoload.php');

$api_instance = new Upcloud\ApiClient\Upcloud\AccountApi();
$config = $api_instance->getConfig();
$config->setUsername('user');
$config->setPassword('password');


try {
  $result = $api_instance->getAccount();
      print_r($result);
} catch (Exception $e) {
  echo 'Exception when calling AccountApi->getAccount: ', $e->getMessage(), PHP_EOL;
}


$firewall_rule = new \Upcloud\ApiClient\Model\FirewallRule();
$firewall_rule->setFamily('IPv4');
$firewall_rule->setSourceAddressStart('123.123.123.0');
$firewall_rule->setAction($firewall_rule::ACTION_DROP);
$firewall_rule->setDirection($firewall_rule::DIRECTION_OUT);
$firewall_rule->setComment('web_client connection rule');
$firewall_rule_request = new \Upcloud\ApiClient\Model\FirewallRuleRequest(); // \Upcloud\ApiClient\Model\FirewallRuleRequest |
$firewall_rule_request->setFirewallRule([$firewall_rule]);
?>
```

So that there is as little as possible to fail. (left the user stuff there to confirm that something works)

Reply:
@PopoSensei  Thanks for the provided info! We'll add the new commit with fixes shortly.

Let us know if you'll still be facing any issues.

Here's an example of how you can create `FirewallRule`:

`$firewall_rule = new \Upcloud\ApiClient\Model\FirewallRule([
  'family' => 'IPv4',
'source_address_start' => '123.123.123.0'
...
]);`

Reply:
Still running into this issue with the latest commit:
excerpt from composer.lock:
```
...
            "name": "upcloudltd/upcloud-php-api",
            "version": "dev-php-version-update",
            "source": {
                "type": "git",
                "url": "https://github.com/DevimaSolutions/upcloud-php-api-fork.git",
                "reference": "55eba3a7aa0ed77ea68a859802c15be71b82bd07"
            },
...
```